### PR TITLE
Catch more c++ exceptions

### PR DESCRIPTION
**Pull request by kdmurray91**
This PR is ready for review. see #1016 for discussion

- [x] Is it mergeable?
- [x] Did it pass the tests?
- [x] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [x] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Is the Copyright year up to date?

### CODE REVIEW

**Comment by mr-c** [comment]
Thanks! Keep 'em coming :-)

**Comment by kdmurray91** [comment]
restest this please jenkins

**Comment by kdmurray91** [comment]
retest this please jenkins

**Comment by kdmurray91** [comment]
Jenkins retest this please you oaf.

**Comment by kdmurray91** [comment]
retest this please jenkins

**Comment by kdmurray91** [comment]
retest this please jenkins

**Comment by mr-c** [comment]
@kdmurray91 FYI, there is now a 'rebuild' button available if you log in to Jenkins. Just go to any build for a particular pull request and look for it on the left side. Feature installed by @luizirber

**Comment by kdmurray91** [comment]
OK, so this has got quite huge. I tried to rebase, and between this and stuff to master since, git's merging algorithm has had some kind of melt-down. Are @mr-c or maybe @luizirber or @camillescott able to help out with the merge? It appears that most conflicts are with @luizirber py3 support branch.

**Comment by luizirber** [comment]
@kdmurray91 I'll take a look, between #978 and #1044 pretty much all the extension module was touched or moved around...

**Comment by luizirber** [comment]
@kdmurray91 take a look on [gotta-catch-them-all-rebased][1], I think I got almost everything. If it is OK, you can just push it over this one.

[1]: https://github.com/dib-lab/khmer/compare/gotta-catch-them-all-rebased?expand=1

**Comment by kdmurray91** [comment]
@luizirber wow, thanks. Yeah those two PRs were both pretty massive. I'll force-push your branch to this, so the PR commentary is maintained.

**Comment by kdmurray91** [comment]
Ready for review @mr-c @luizirber @camillescott @ctb. The test failure is due to conditional coverage, which is 0.08 percent off :grimacing:

**Comment by kdmurray91** [comment]
Remaining questions:

- [ ] Does the `CatchIOErrors` context manager in `scripts/normalize-by-median.py` need updating to OSError?
- [ ]  There's a few scripts in `sandbox/` which catch IOError, seemingly both from python's `open()` and `_khmer.cc`. So these probably require `except (IOError, OSError) as exc:`?

File: tests/test_scripts.py [hunk 1]
```diff
@@ -591,6 +591,462 @@ def test_filter_stoptags_fq():
     assert 'seq 1::BAR' in names
 
 
+def test_normalize_by_median_indent():
```
Comment (kdmurray91) [inline tests/test_scripts.py]: @luizirber , is this block of added tests an extraneous inclusion? I didn't write them...

**Comment by kdmurray91** [comment]
jenkins retest this

**Comment by kdmurray91** [comment]
jenkins retest this please!

File: tests/test_scripts.py (hunk 1 above)
Comment (luizirber) [inline tests/test_scripts.py]: Yes, it is. Sorry!

Comment (kdmurray91) [inline tests/test_scripts.py]: Cool man, no worries. I'll remove them :)

**Comment by bocajnotnef** [comment]
In theory, https://github.com/dib-lab/khmer/blob/master/tests/test_normalize_by_median.py#L76 should test for the error handling. If it's passing then we should be A-OK--though from glancing through the changes I'd expect it to be breaking... Strange.

**Comment by mr-c** [comment]
@kdmurray91 Yep, we need to add OSError as appropriate.

**Comment by kdmurray91** [comment]
@bocajnotnef My thoughts exactly. Hmmm. Will poke around and see if I can make it fail.
@mr-c  Done!

**Comment by kdmurray91** [comment]
@bocajnotnef No idea what's going on but I can't work out why it isn't failing. Are you happy for me to just add `OSError` to the `except:` block?

**Comment by kdmurray91** [comment]
Also, the CI status seems to have been dropped from the PR, so I'll ask jenkins  to retest this please and see what happens.

**Comment by bocajnotnef** [comment]
@kdmurray91 I thought adding the `OSError` to the `except` was the original plan since it looked like all/most of the `IOErrors` had been changed to OSs.

**Comment by kdmurray91** [comment]
@bocajnotnef  Yep, it was! 

The failing CI status is due to code coverage alone, I think. So I think we're good for CR @bocajnotnef @luizirber @mr-c

**Comment by kdmurray91** [comment]
retest this please jenkins

**Comment by kdmurray91** [comment]
I'm assuming this is higher priority than #1018, so I'll rebase to make this mergable first.

**Comment by mr-c** [comment]
I am a huge fan of this PR. LGTM. Please make mergeable and merge this after Jenkins approves.

**Comment by kdmurray91** [comment]
@mr-c Thanks, am doing!

**Comment by kdmurray91** [comment]
also, jenkins, retest this please

**Comment by kdmurray91** [comment]
I'm also going to fix a few unrelated python2-isms so I can test with py3.4.

**Comment by kdmurray91** [comment]
Ok, so test appear to fail solely due to coverage. On my machine, for python3 to work, a default must be explicitly set in `sandbox/collect-reads.py`:

```diff
diff --git a/sandbox/collect-reads.py b/sandbox/collect-reads.py
index f02c0ea..380a297 100755
--- a/sandbox/collect-reads.py
+++ b/sandbox/collect-reads.py
@@ -28,6 +28,9 @@ import argparse
 import screed
 
 
+DEFAULT_COVERAGE = 20
+
+
 def output_single(read):
     if hasattr(read, 'quality'):
         return "@%s\n%s\n+\n%s\n" % (read.name, read.sequence, read.quality)
@@ -54,7 +57,7 @@ def get_parser():
                         "sequence files.")
     parser.add_argument('--report-total-kmers', '-t', action='store_true',
                         help="Prints the total number of k-mers to stderr")
-    parser.add_argument('-C', '--coverage', type=int,
+    parser.add_argument('-C', '--coverage', default=DEFAULT_COVERAGE, type=int,
                         help='Collect reads until this coverage, then exit.')
     parser.add_argument('-o', '--output', type=argparse.FileType('w'),
                         help='Write collect reads into this file.')
```

**Comment by kdmurray91** [comment]
OK, so I've decided to add tests for the issues that triggered this behemoth, #1016 and #1015.

**Comment by kdmurray91** [comment]
retest this please jenkins

File: tests/test_counting_hash.py [hunk 2]
```diff
@@ -974,7 +985,9 @@ def test_get_ksize():
 
 def test_get_hashsizes():
     kh = khmer.CountingHash(22, 100, 4)
-    assert kh.hashsizes() == [97L, 89L, 83L, 79L], kh.hashsizes()
+    expected = utils.longify([97, 89, 83, 79])  # Py3 hack.
```
Comment (mr-c) [inline tests/test_counting_hash.py]: Does the `Py3 hack` comment mean that this can be removed should we go Py3 only?

File: lib/trace_logger.cc [hunk 3]
```diff
@@ -40,7 +40,11 @@ TraceLogger( uint8_t const level, char const * const file_name_format, ... )
 
     _stream_handle = fopen( tfile_name, "w" );
     if (NULL == _stream_handle) {
-        throw InvalidStreamBuffer( );
+        std::ostringstream msg;
```
Comment (mr-c) [inline lib/trace_logger.cc]: Thanks. Does this mean that you are using the trace code in your work?

File: tests/test_counting_hash.py (hunk 2 above)
Comment (kdmurray91) [inline tests/test_counting_hash.py]: It's a work-around for not having long literals, and needing them in py2. Is this an acceptable solution?

File: lib/trace_logger.cc (hunk 3 above)
Comment (kdmurray91) [inline lib/trace_logger.cc]: Nope, just got caught in my `git grep` marathon. The `InvalidStreamBuffer` exception class doesn't exist any longer.

File: tests/test_counting_hash.py (hunk 2 above)
Comment (mr-c) [inline tests/test_counting_hash.py]: Fine solution, yep! An actionable comment would be helpful -> "Remove once Py2 is dropped" or somesuch.

File: lib/trace_logger.cc (hunk 3 above)
Comment (mr-c) [inline lib/trace_logger.cc]: Coo', good to know

**Comment by mr-c** [comment]
Still looking good to me :-)

**Comment by kdmurray91** [comment]
And I think I'm done. Time for the big green button? Any missed cases should be picked up by the pre-release testing right? I'll give priority to cleaning up loose ends pre 2.0 release.

**Comment by mr-c** [comment]
Yep, thanks!

**Pull request merged**